src/main/java/org/apache/flink/scheduler/benchmark/SchedulerBenchmarkUtils.java

```diff
@@ -0,0 +1,269 @@
+			predicateResult = predicate.test(list);
+
+			if (!predicateResult) {
+				try {
+					Thread.sleep(2L);
+				} catch (InterruptedException ignored) {
+					Thread.currentThread().interrupt();
+				}
+			}
+		} while (!predicateResult && deadline.hasTimeLeft());
+
+		if (!predicateResult) {
+			throw new TimeoutException(String.format(
+					"List no fulfilled in time, expected %d, actual %d.",
+					length,
+					list.size()));
+		}
+	}
+
+	public static void verifyListSize(
+			Collection<?> list,
+			int length) {
+		if (list.size() < length) {
+			throw new RuntimeException(String.format(
+					"Size of the list mismatch, expected %d, actual %d.",
+					length,
+					list.size()));
+		}
+	}
+
+	public static void deployTasks(
+			ExecutionGraph executionGraph,
+			JobVertexID jobVertexID,
+			TestingLogicalSlotBuilder slotBuilder,
+			boolean sendScheduleOrUpdateConsumersMessage) throws JobException, ExecutionException, InterruptedException {
+
+		for (ExecutionVertex vertex : executionGraph.getJobVertex(jobVertexID).getTaskVertices()) {
+			LogicalSlot slot = slotBuilder.createTestingLogicalSlot();
+			Execution execution = vertex.getCurrentExecutionAttempt();
+			execution.registerProducedPartitions(
+					slot.getTaskManagerLocation(),
+					sendScheduleOrUpdateConsumersMessage).get();
+			assignResourceAndDeploy(vertex, slot);
+		}
+	}
+
+	public static void deployAllTasks(
+			ExecutionGraph executionGraph,
+			TestingLogicalSlotBuilder slotBuilder) throws JobException, ExecutionException, InterruptedException {
+
+		for (ExecutionVertex vertex : executionGraph.getAllExecutionVertices()) {
+			LogicalSlot slot = slotBuilder.createTestingLogicalSlot();
+			vertex.getCurrentExecutionAttempt().registerProducedPartitions(slot.getTaskManagerLocation(), true).get();
+			assignResourceAndDeploy(vertex, slot);
+		}
+	}
+
+	private static void assignResourceAndDeploy(ExecutionVertex vertex, LogicalSlot slot) throws JobException {
+		vertex.tryAssignResource(slot);
+		vertex.deploy();
+	}
+
+	public static void transitionTaskStatus(
+			ExecutionGraph executionGraph,
+			JobVertexID jobVertexID,
+			ExecutionState state) {
+
+		for (ExecutionVertex vertex : executionGraph
+				.getJobVertex(jobVertexID)
+				.getTaskVertices()) {
+			executionGraph.updateState(new TaskExecutionStateTransition(new TaskExecutionState(
+					executionGraph.getJobID(),
+					vertex.getCurrentExecutionAttempt().getAttemptId(),
+					state)));
+		}
+	}
+
+	public static void transitionTaskStatus(
+			DefaultScheduler scheduler,
+			AccessExecutionJobVertex vertex,
+			int subtask,
+			ExecutionState executionState) {
+
+		final ExecutionAttemptID attemptId = vertex.getTaskVertices()[subtask]
+				.getCurrentExecutionAttempt()
+				.getAttemptId();
+		scheduler.updateTaskExecutionState(
+				new TaskExecutionState(
+						scheduler.getExecutionGraph().getJobID(),
+						attemptId,
+						executionState));
+	}
+
+	public static void transitionDownstreamTasksToRunning(
```

Comment:
This method is a bit weird and hacky because it has assumption on the topology. 
Maybe we can split it into 3 invocations and conduct them in benchmarks: deploy(sourceVertex) -> finish(sourceVertex) -> deploy(sinkVertex).

Note that it's valid to transition a DEPLOYING task to FINISHED, so we do not need to transition them to RUNNING.

Reply:
Totally agreed, this method seems to be unnecessary. I've removed this method.